File: matlab/ch10/bnbopt.m
function [xopt,fopt,iter] = bnbopt(A,b,c,nl,ng,ne,ibd)
% bnbopt.m: branch and bound method for mixed inter minimization problem
% Inputs:
% A: coefficient matrix for constraints
% b: right hand side of constraints (row vector)
% c: coefficient vector for objective function (row vector)
% nl: number of <= constraints
% ng: number of >= constraints
% ne: number of = constraints
% ibd: variable index vector
% Outputs:
% xopt: optimal point (0 or 1)
% fopt: function value at the optimal point (=f(xopt))
% iter: number of iterations
critd = 1e-4; critf = 5e-2; kmax = 1e4; ksol = 0; kf = 0; fmin = 1e30;
nv = length(c); nbd = length(ibd);
for j = 1:length(b)
if (b(j) < 0), disp('b(j) must be non-negative.'); return; end
end
for k = 1:nv, vtype(k) = 1; end
for k = 1:nbd, vtype(ibd(k)) = 2; end
for j = 1:nv, kvar(j) = j; end
iter = 0; kconv = 0;
while (1)
iter = iter + 1;
if (iter > kmax), disp('Maximum possible iterations exceeded'); break; end
ktrac = 0; [f,x,kflag] = linpm(nv,nl,ng,ne,kf,kvar,A,b,c);
if (nbd == 0)
if (kflag > 0), disp('No feasible solution.'); return;
else, fmin = f; xmin = x; break; end
end
jc = 0;
while (2)
jc = jc+1; if (jc == 2), break; end
if (kflag > 0)
if (kf == 0), disp('No feasible solution.'); return; end
ktrac = 1; break; % backtrack operation
else
if (iter == 1), flb = f; end % best low bound
if (f >= fmin), ktrac = 1; break; end
vmax = 0;
for j = 1:nv
if(vtype(j) == 2)
diffx = abs(x(j)-round(x(j)));
if (diffx > vmax), vmax = diffx; jv = j; end
end
end
if (vmax < critd)
ksol = 1;
if (f < fmin), fmin = f; for j = 1:nv, xmin(j) = x(j); end; end
gapf = abs(flb - fmin)/abs(flb);
if (gapf <= critf), kconv = 1; break; end
ktrac = 1; % backtrack operation
break
else
kf = kf + 1; jt = kvar(kf); kvar(kf) = kvar(jv); kvar(jv) = jt;
end
end
end
while (ktrac == 1)
if (kvar(kf) > 0), kvar(kf) = -kvar(kf); break;
else
kvar(kf) = -kvar(kf); kf = kf - 1;
if (kf == 0)
if(ksol == 0), disp('No feasible solution.'); return; end
kconv = 1; break;
end
end
end
if (kconv == 1), break; end
end
xopt = xmin; fopt = fmin;
end
function [f,x,kflag] = linpm(nv,nl,ng,ne,kf,kvar,A,B,C)
if (kf > 0)
idn = [1:nv];
for k = 1:kf, jk = abs(kvar(k)); idn(jk) = 0; end
difnv = nv - kf; nej = ne; Aj = A; Bj = B; f0 = 0; cj = 0; jm = 0;
for k = 1:nv
if idn(k) > 0, cj = cj+1; Ck(cj) = C(k);
else
jm = jm+1;
if (kvar(jm) > 0), cm = 1; else, cm = 0;
end
f0 = f0 + C(k)*cm;
end
end
for k = 1:nl, conk(k) = -1; end
for k = 1:ng, conk(nl+k) = 1; end
for k = 1:nl+ng+ne
cj = 0; jm = 0;
for j = 1:nv
if (idn(j)) > 0, cj = cj + 1; Aj(k,cj) = A(k,j);
else
jm = jm+1;
if (kvar(jm) > 0), cm = 1; else, cm = 0; end
Bj(k) = Bj(k) - A(k,j)*cm;
end
end
if (k <= nl+ng & Bj(k) < 0), conk(k) = -conk(k); Aj(k,:) = -Aj(k,:); Bj(k) = -Bj
(k); end
end
nlk = 0; ngk = 0;
for k = 1:nl+ng
if (conk(k) == -1), nlk = nlk + 1;
elseif (conk(k)==1), ngk = ngk + 1;
end
end
[ids,xk] = sort(conk); A = Aj; B = Bj;
for k = 1:nl+ng, A(k,:) = Aj(xk(k),:); B(k) = Bj(xk(k)); end
nk = nlk + ngk + nej; Aj = A(1:nk,1:difnv); Bj = B(1:nk);
else
difnv = nv; nlk = nl; ngk = ng; nej = ne; Aj = A; Bj = B; Ck = C;
end
[fk,xs,kflag] = linsimx(difnv,nlk,ngk,nej,Aj,Bj,Ck);
if (kflag > 0), x = xs; f = fk; return; end
if (kf > 0)
cj = 0; jm = 0;
for j = 1:nv
if (idn(j)) > 0. cj = cj+1; x(j) = xs(cj);
else
jm = jm+1;
if (kvar(jm)>0), cm = 1; else, cm = 0; end
x(j) = cm;
end
end
f = fk + f0;
else
f = fk; x = xs; return;
end
end

function [f,x,kflag] = linsimx(nv,nl,ng,ne,A,B,C)
% linear programming by simplex method
mnp = 20; bigm = 0; kflag = 0;
nc = nv + nl + ne + 2*ng; nr = nl + ne + ng + 1;
% Initialization
Aj(1:nr,1:nc) = 0; Bj(1:nr) = 0; Aj(1:nr-1,1:nv) = A(1:nr-1,1:nv);
Aj(nr,1:nv) = C(1:nv); Bj(1:nr-1) = B(1:nr-1); A = Aj; B = Bj;
for j = 1:nv, A(nr,j) = C(j); end
for j = 1:nv, bigm = bigm + mnp*abs(A(nr,j)); end
if (nl > 0) % slack variables
for k = 1:nl, A(k,nv+k) = 1; Bs(k) = nv + k; end
end
if (ng > 0)
for k = 1:ng
A(nl+k,nv+nl+k) = -1; A(nl+k, nv+nl+ng+k) = 1; Bs(nl+k) = nv + nl + ng + k;
end
end
if (ne > 0)
for k = 1:ne, A(nl+ng+k,nv+nl+2*ng+k) = 1; Bs(nl+ng+k) = nv + nl + 2*ng + k; end
end
mge = ng + ne;
if (mge > 0), for k = 1:mge, A(nr,nv+nl+ng+k) = bigm; end
end
if (mge > 0) % Remove artificial variables
for k = 1:mge
C = A(nr,nv+nl+ng+k); for j = 1:nc, A(nr,j) = A(nr,j) - C*A(nl+k,j); end
B(nr) = B(nr) - C*B(nl+k);
end
end

% Simplex method
while (1)
jflag = 0;
for j = 1:nc, if (A(nr,j) < 0), jflag = 1; break; end; end
if (jflag == 0), break; end
C = bigm;
for j = 1:nc, if (A(nr,j) < C), C = A(nr,j); idv = j; end; end
jn = 0; jk = 0;
for k = 1:nr-1
if (A(k,idv) > 0)
jk = jk + 1; Dm = B(k)/(A(k,idv) + 1e-10);
if (jk == 1), C = Dm; jp = k;
else, if (Dm < C), C = Dm; jp = k; end
end
jn = 1;
end
end
Bs(jp) = idv; if (jn == 0), kflag = 1; f = 0; x = 0; return; end
Dm = 1/A(jp,idv); B(jp) = Dm*B(jp);
for j = 1:nc, A(jp,j) = Dm*A(jp,j); end
for k = 1:nr
if (k ~= jp)
Em = A(k,idv); for j = 1:nc, A(k,j) = A(k,j) - Em*A(jp,j); end
B(k) = B(k) - Em*B(jp);
end
end
end
x(1:nv) = 0; nt = nv + nl + ng;
for k = 1:nr-1, if (Bs(k) > nt), kflag = 1; f = 0; x = 0; return; end; end
for k = 1:nv
for j = 1:nr-1, if (Bs(j) == k), x(k) = B(j); break; end; end
end
f = B(nr); f = -f; % minimization
end

File: python/ch10/bnbopt.py
import numpy as np

def bnbopt(A, b, c, nl, ng, ne, ibd):
    """
    Branch and Bound method for mixed integer minimization.
    """
    critd = 1e-4
    critf = 5e-2
    kmax = 1e4
    ksol = 0
    kf = 0
    fmin = 1e30
    
    A = np.array(A, dtype=float)
    b = np.array(b, dtype=float)
    c = np.array(c, dtype=float)
    
    nv = len(c)
    nbd = len(ibd)
    
    # Validation
    if np.any(b < 0):
        print('b(j) must be non-negative.')
        return None, None, 0

    vtype = np.ones(nv)
    for idx in ibd:
        vtype[idx-1] = 2 # MATLAB 1-based to 0-based
        
    kvar = np.array(list(range(1, nv + 1))) # 1-based indices for logic
    iter_count = 0
    kconv = 0
    xmin = np.zeros(nv)

    while True:
        iter_count += 1
        if iter_count > kmax:
            print('Maximum possible iterations exceeded')
            break
            
        ktrac = 0
        # linpm 호출
        f, x, kflag = linpm(nv, nl, ng, ne, kf, kvar, A, b, c)
        
        if nbd == 0:
            if kflag > 0:
                print('No feasible solution.')
                return None, None, iter_count
            else:
                fmin = f
                xmin = x
                break
        
        jc = 0
        while True:
            jc += 1
            if jc == 2: break
            
            if kflag > 0:
                if kf == 0:
                    print('No feasible solution.')
                    return None, None, iter_count
                ktrac = 1
                break # backtrack
            else:
                if iter_count == 1:
                    flb = f # best lower bound
                
                if f >= fmin:
                    ktrac = 1
                    break
                
                vmax = 0
                jv = -1
                for j in range(nv):
                    if vtype[j] == 2:
                        diffx = abs(x[j] - round(x[j]))
                        if diffx > vmax:
                            vmax = diffx
                            jv = j + 1 # 1-based index
                
                if vmax < critd:
                    ksol = 1
                    if f < fmin:
                        fmin = f
                        xmin = np.copy(x)
                    
                    gapf = abs(flb - fmin) / abs(flb) if abs(flb) > 1e-10 else 0
                    if gapf <= critf:
                        kconv = 1
                        break
                    ktrac = 1
                    break
                else:
                    kf += 1
                    # Swap logic
                    idx_kf = kf - 1
                    idx_jv = np.where(np.abs(kvar) == jv)[0][0]
                    kvar[idx_kf], kvar[idx_jv] = kvar[idx_jv], kvar[idx_kf]

        while ktrac == 1:
            if kvar[kf-1] > 0:
                kvar[kf-1] = -kvar[kf-1]
                break
            else:
                kvar[kf-1] = abs(kvar[kf-1])
                kf -= 1
                if kf == 0:
                    if ksol == 0:
                        print('No feasible solution.')
                        return None, None, iter_count
                    kconv = 1
                    break
        
        if kconv == 1: break

    return xmin, fmin, iter_count

def linpm(nv, nl, ng, ne, kf, kvar, A, B, C):
    """
    Sub-function to prepare the sub-problem for the simplex solver.
    """
    if kf > 0:
        idn = np.ones(nv, dtype=bool)
        for k in range(kf):
            jk = abs(kvar[k]) - 1
            idn[jk] = False
        
        difnv = nv - kf
        f0 = 0
        
        # Determine fixed values and objective offset
        for k in range(kf):
            val = 1 if kvar[k] > 0 else 0
            f0 += C[abs(kvar[k])-1] * val
            
        # Build reduced problem
        Ck = C[idn]
        Aj = A[:, idn]
        Bj = B.copy()
        
        for k in range(len(B)):
            for m in range(kf):
                val = 1 if kvar[m] > 0 else 0
                Bj[k] -= A[k, abs(kvar[m])-1] * val
        
        # Simple constraint type tracking (simplified for conversion)
        # MATLAB logic for conk needs to handle original indices
        conk = np.zeros(nl + ng + ne)
        conk[:nl] = -1
        conk[nl:nl+ng] = 1
        
        for k in range(nl + ng + ne):
            if k < (nl + ng) and Bj[k] < 0:
                conk[k] = -conk[k]
                Aj[k, :] = -Aj[k, :]
                Bj[k] = -Bj[k]
        
        # Sort and filter
        # Note: This part depends on linsimx implementation
        fk, xs, kflag = linsimx(difnv, nl, ng, ne, Aj, Bj, Ck)
        
        if kflag > 0:
            return 0, None, kflag
        
        x = np.zeros(nv)
        xs_idx = 0
        for j in range(nv):
            if idn[j]:
                x[j] = xs[xs_idx]
                xs_idx += 1
            else:
                # Find which kvar entry corresponds to this variable
                for m in range(kf):
                    if abs(kvar[m]) == j + 1:
                        x[j] = 1 if kvar[m] > 0 else 0
        f = fk + f0
        return f, x, kflag
    else:
        return linsimx(nv, nl, ng, ne, A, B, C)

def linsimx(nv, nl, ng, ne, A, B, C):
    """
    Placeholder for the Simplex solver. 
    You need to implement or link a Simplex solver here.
    """
    # 원본 코드에는 linsimx의 구현이 없으므로, 
    # scipy.optimize.linprog 등을 활용하거나 별도 구현이 필요합니다.
    from scipy.optimize import linprog
    
    # scipy linprog는 기본적으로 <= 제약을 사용함
    # nl: <=, ng: >=, ne: =
    A_ub = []
    b_ub = []
    A_eq = []
    b_eq = []
    
    if nl > 0:
        A_ub.extend(A[:nl])
        b_ub.extend(B[:nl])
    if ng > 0:
        A_ub.extend(-A[nl:nl+ng])
        b_ub.extend(-B[nl:nl+ng])
    if ne > 0:
        A_eq.extend(A[nl+ng:])
        b_eq.extend(B[nl+ng:])
        
    res = linprog(C, A_ub=A_ub if A_ub else None, b_ub=b_ub if b_ub else None,
                  A_eq=A_eq if A_eq else None, b_eq=b_eq if b_eq else None, 
                  bounds=(0, 1), method='highs')
    
    if res.success:
        return res.fun, res.x, 0
    else:
        return 0, None, 1
        
import numpy as np

def linsimx(nv, nl, ng, ne, A_in, B_in, C_in):
    """
    Linear programming by simplex method.
    """
    mnp = 20
    bigm = 0
    kflag = 0
    
    nc = nv + nl + ne + 2 * ng
    nr = nl + ne + ng + 1
    
    # Initialization
    # MATLAB: Aj(1:nr, 1:nc) = 0
    Aj = np.zeros((nr, nc))
    Bj = np.zeros(nr)
    
    # Data mapping
    Aj[:nr-1, :nv] = A_in[:nr-1, :nv]
    Aj[nr-1, :nv] = C_in
    Bj[:nr-1] = B_in[:nr-1]
    
    A = Aj
    B = Bj
    
    # Calculate BigM
    for j in range(nv):
        bigm += mnp * abs(A[nr-1, j])
        
    Bs = np.zeros(nr, dtype=int)
    
    # Slack variables (nl)
    if nl > 0:
        for k in range(nl):
            A[k, nv + k] = 1
            Bs[k] = nv + k + 1 # 1-based indexing
            
    # Surplus and Artificial variables (ng)
    if ng > 0:
        for k in range(ng):
            idx = nl + k
            A[idx, nv + nl + k] = -1
            A[idx, nv + nl + ng + k] = 1
            Bs[idx] = nv + nl + ng + k + 1
            
    # Artificial variables (ne)
    if ne > 0:
        for k in range(ne):
            idx = nl + ng + k
            A[idx, nv + nl + 2 * ng + k] = 1
            Bs[idx] = nv + nl + 2 * ng + k + 1
            
    # BigM inclusion
    mge = ng + ne
    if mge > 0:
        for k in range(mge):
            A[nr-1, nv + nl + ng + k] = bigm
            
    # Remove artificial variables
    if mge > 0:
        for k in range(mge):
            row_idx = nl + k
            C = A[nr-1, nv + nl + ng + k]
            A[nr-1, :] -= C * A[row_idx, :]
            B[nr-1] -= C * B[row_idx]
            
    # Simplex iteration
    while True:
        jflag = 0
        # Find pivot column
        if np.any(A[nr-1, :nc] < 0):
            jflag = 1
        
        if jflag == 0: break
        
        # Select idv (most negative)
        idv = np.argmin(A[nr-1, :nc])
        
        # Find pivot row
        jn = 0
        jk = 0
        jp = -1
        min_ratio = float('inf')
        
        for k in range(nr - 1):
            if A[k, idv] > 0:
                jk += 1
                Dm = B[k] / (A[k, idv] + 1e-10)
                if jk == 1 or Dm < min_ratio:
                    min_ratio = Dm
                    jp = k
                    jn = 1
        
        if jn == 0:
            return 0, 0, 1
        
        Bs[jp] = idv + 1 # Store 1-based index
        
        # Pivot operation
        Dm = 1.0 / A[jp, idv]
        B[jp] *= Dm
        A[jp, :] *= Dm
        
        for k in range(nr):
            if k != jp:
                Em = A[k, idv]
                A[k, :] -= Em * A[jp, :]
                B[k] -= Em * B[jp]
                
    # Result extraction
    x = np.zeros(nv)
    nt = nv + nl + ng
    
    # Check for feasibility (artificial variables should not be in basis)
    for k in range(nr - 1):
        if Bs[k] > nt:
            return 0, 0, 1
            
    for k in range(1, nv + 1):
        for j in range(nr - 1):
            if Bs[j] == k:
                x[k-1] = B[j]
                break
                
    f = -B[nr-1] # Minimize
    return f, x, 0